### PR TITLE
# 161476173 login user

### PR DESCRIPTION
**Pull request by wiztemple**
### What does this PR do?
Logs in a registered user to the application

### Description of Tasks to be completed
- create a method to login user
- generate token for user

### How should this be manually tested?
- git clone the repository
- cd into the the repo folder
- run git checkout ft-users-receive-jwt-successful-161476173

Open up postman and hit the following localhost:5000/api/v1/users/signin route with the json input
```
{
   email: "wisdom@gmail.com",
   password: "weirdcode",
}
```

Run npm test

### What are the relevant Pivotal Tracker Stories?
- [#161476173](https://www.pivotaltracker.com/story/show/161476173)

### Screenshots(if appropriate)
screen shot 2018-10-30 at 10 31 51 am

screenshot 2018-10-29 at 7 07 08 pm

screen shot 2018-10-30 at 7 24 11 am

### CODE REVIEW

**Comment by PriscillaSam** [comment]
This looks alright.

File: server/controllers/user.js [hunk 1]
```diff
@@ -49,6 +49,42 @@ class User {
...
+        response,
+        200,
+        `Hello ${findUser.username}, Welcome Back!`,
+        token
+      );
+    }
+    return badHttpResponse(
+      response,
+      404,
+      'User does not exist'
```
Comment (TheoOkafor) [inline server/controllers/user.js]: I think the error message should also inform the user about the actions to take. Consider changing this to ```"This account does not exist. Consider sign up."```

File: server/tests/controllerTests/user.test.js [hunk 2]
```diff
@@ -36,3 +37,62 @@ describe('POST api/users/signup', () => {
...
+      .post(signinRoute)
+      .send(sull);
+    expect(response.status).to.equal(200);
+    expect(response.body).to.be.an('object');
+    expect(response.body.message).to.deep.equals('Hello i_amsull, Welcome Back!');
+  });
+  it('should not sign in user without password', async () => {
+    const response = await chai.request(app)
+      .post(signinRoute)
+      .send({
```
Comment (TheoOkafor) [inline server/tests/controllerTests/user.test.js]: Consider storing this in the mockData.js file created by Priscilla

File: server/tests/controllerTests/user.test.js [hunk 3]
```diff
@@ -36,3 +37,62 @@ describe('POST api/users/signup', () => {
...
+      });
+    expect(response).to.have.status(400);
+    expect(response.body).to.be.an('object');
+    expect(response.body.message).to.be.deep.equals('Invalid input');
+    expect(response.body.error.problem.email).to.be.deep.equals('The Email provided is invalid. ');
+  });
+  it('should not sign in user with an incorrect password', async () => {
+    const response = await chai.request(app)
+      .post(signinRoute)
+      .send({
```
Comment (TheoOkafor) [inline server/tests/controllerTests/user.test.js]: This one too and the other test data.

File: server/controllers/user.js [hunk 4]
```diff
@@ -49,6 +49,42 @@ class User {
...
+          response,
+          403,
+          'incorrect password, please try again'
+        );
+      }
+      const token = generateToken(findUser.id);
+      return goodHttpResponse(
+        response,
+        200,
+        `Hello ${findUser.username}, Welcome Back!`,
```
Comment (TheoOkafor) [inline server/controllers/user.js:78]: Nice message Gents. I like it 👍

File: server/controllers/user.js [hunk 5]
```diff
@@ -49,6 +49,42 @@ class User {
...
+        token
```
Comment (PriscillaSam) [inline server/controllers/user.js]: @wiztemple @ucheg6 please put the token in an object like this `{ token }` so that the data object from the response has the attribute as `token` and not `output`.

File: server/controllers/user.js [hunk 6]
```diff
@@ -49,6 +49,42 @@ class User {
...
+      email,
+      password,
+    } = request.body;
+    const findUser = await userRepo.getUserByEmail(email);
+    if (findUser) {
+      const validatePassword = await compare(password, findUser.password);
+      if (validatePassword === false) {
+        return badHttpResponse(
+          response,
+          403,
```
Comment (seyi-adeleke) [inline server/controllers/user.js]: This should be a 401

File: server/middlewares/validations.js [hunk 7]
```diff
@@ -17,4 +17,14 @@ const inputValidator = (request, response, next) => {
   next();
 };
 
-export default inputValidator;
+const signinValidation = (request, response, next) => {
+  const { isValid, problem } = checkMissingFields(request.body, [
+    'email',
+    'password',
+  ]);
+  if (isValid === false) {
```
Comment (seyi-adeleke) [inline server/middlewares/validations.js]: `!isValid` for brevity

File: server/routes/users.js [hunk 8]
```diff
@@ -1,9 +1,10 @@
 import { Router } from 'express';
 import User from '../controllers/user';
-import inputValidator from '../middlewares/validations';
+import { inputValidator, signinValidation } from '../middlewares/validations';
 
 const router = new Router();
 
 router.post('/users/signup', inputValidator, User.signup);
```
Comment (seyi-adeleke) [inline server/routes/users.js]: `inputValidator` should be `signupValidator` then

**Comment by ucheg6** [comment]
@seyi-adeleke all your feedback has been addressed

**Review by seyi-adeleke — APPROVED** [review]

File: package.json [hunk 9]
```diff
@@ -42,6 +42,7 @@
     "babel-plugin-add-module-exports": "^1.0.0",
     "babel-plugin-istanbul": "^5.1.0",
     "babel-plugin-transform-runtime": "^6.23.0",
+    "babel-polyfill": "^6.26.0",
```
Comment (seyi-adeleke) [inline package.json]: Why did you add this package?

Comment (ucheg6) [reply package.json]: I checked in with Priscilla and Theo and they said they put it there to enable the async/await syntax

Comment (ucheg6) [reply package.json]: They also said that the package is not needed again because of the presence of babel-plugin-transform-runtime in .babelrc. So we agreed I should remove it.
